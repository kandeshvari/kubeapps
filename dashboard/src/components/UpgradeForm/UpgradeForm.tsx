import { RouterAction } from "connected-react-router";
import { JSONSchema4 } from "json-schema";
import * as React from "react";

import { IChartState, IChartVersion } from "../../shared/types";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";

export interface IUpgradeFormProps {
  appCurrentVersion: string;
  appCurrentValues?: string;
  chartName: string;
  namespace: string;
  releaseName: string;
  repo: string;
  error: Error | undefined;
  selected: IChartState["selected"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  fetchChartVersions: (id: string) => Promise<IChartVersion[]>;
  getChartVersion: (id: string, chartVersion: string) => void;
}

interface IUpgradeFormState {
  appValues: string;
  valuesModified: boolean;
  isDeploying: boolean;
}

class UpgradeForm extends React.Component<IUpgradeFormProps, IUpgradeFormState> {
  public state: IUpgradeFormState = {
    appValues: this.props.appCurrentValues || "",
    isDeploying: false,
    valuesModified: false,
  };

  public render() {
    const { namespace, releaseName, error } = this.props;
    if (error) {
      return (
        <ErrorSelector error={error} namespace={namespace} action="update" resource={releaseName} />
      );
    }

    const chartID = `${this.props.repo}/${this.props.chartName}`;
    return (
      <form className="container padding-b-bigger" onSubmit={this.handleDeploy}>
        <div className="row">
          <div className="col-12">
            <h2>{`${this.props.releaseName} (${chartID})`}</h2>
          </div>
          <div className="col-8">
            <DeploymentFormBody
              chartID={chartID}
              chartVersion={this.props.appCurrentVersion}
              originalValues={this.props.appCurrentValues}
              namespace={this.props.namespace}
              releaseName={this.props.releaseName}
              selected={this.props.selected}
              push={this.props.push}
              goBack={this.props.goBack}
              fetchChartVersions={this.props.fetchChartVersions}
              getChartVersion={this.props.getChartVersion}
              setValues={this.handleValuesChange}
              appValues={this.state.appValues}
              valuesModified={this.state.valuesModified}
              setValuesModified={this.setValuesModified}
            />
          </div>
        </div>
      </form>
    );
  }

  public setValuesModified = () => {
    this.setState({ valuesModified: true });
  };

  public handleValuesChange = (value: string) => {
    this.setState({ appValues: value });
  };

  public handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { selected, push, upgradeApp, releaseName, namespace } = this.props;
    const { appValues } = this.state;

    this.setState({ isDeploying: true });
    if (selected.version) {
      const deployed = await upgradeApp(
        selected.version,
        releaseName,
        namespace,
        appValues,
        selected.schema,
      );
      this.setState({ isDeploying: false });
      if (deployed) {
        push(`/apps/ns/${namespace}/${releaseName}`);
      }
    }
  };
}

export default UpgradeForm;
