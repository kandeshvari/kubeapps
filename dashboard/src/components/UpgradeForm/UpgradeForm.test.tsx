import { mount, shallow } from "enzyme";
import * as React from "react";

import { IChartState, IChartVersion, UnprocessableEntity } from "../../shared/types";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import UpgradeForm, { IUpgradeFormProps } from "./UpgradeForm";

const defaultProps = {
  appCurrentVersion: "1.0.0",
  appCurrentValues: "foo: bar",
  chartName: "my-chart",
  namespace: "default",
  releaseName: "my-release",
  repo: "my-repo",
  selected: {} as IChartState["selected"],
  upgradeApp: jest.fn(),
  push: jest.fn(),
  goBack: jest.fn(),
  fetchChartVersions: jest.fn(),
  getChartVersion: jest.fn(),
  error: undefined,
} as IUpgradeFormProps;

const versions = [{ id: "foo", attributes: { version: "1.2.3" } }] as IChartVersion[];

describe("renders an error", () => {
  it("renders a custom error if the deployment failed", () => {
    const wrapper = shallow(
      <UpgradeForm
        {...defaultProps}
        selected={
          {
            version: { attributes: {} },
            versions: [{ id: "foo", attributes: {} }],
          } as IChartState["selected"]
        }
        error={new UnprocessableEntity("wrong format!")}
      />,
    );
    wrapper.setState({ latestSubmittedReleaseName: "my-app" });
    expect(wrapper.find(ErrorSelector).exists()).toBe(true);
    expect(wrapper.find(ErrorSelector).html()).toContain(
      "Sorry! Something went wrong processing my-release",
    );
    expect(wrapper.find(ErrorSelector).html()).toContain("wrong format!");
  });
});

it("renders the full UpgradeForm", () => {
  const wrapper = shallow(
    <UpgradeForm {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );
  expect(wrapper).toMatchSnapshot();
});

it("forwards the appValues when modified", () => {
  const wrapper = shallow(<UpgradeForm {...defaultProps} />);
  const handleValuesChange: (v: string) => void = wrapper
    .find(DeploymentFormBody)
    .prop("setValues");
  handleValuesChange("foo: bar");

  expect(wrapper.state("appValues")).toBe("foo: bar");
  expect(wrapper.find(DeploymentFormBody).prop("appValues")).toBe("foo: bar");
});

it("forwards the valuesModifed property", () => {
  const wrapper = shallow(<UpgradeForm {...defaultProps} />);
  const handleValuesModified: (v: string) => void = wrapper
    .find(DeploymentFormBody)
    .prop("setValuesModified");
  handleValuesModified("foo: bar");

  expect(wrapper.state("valuesModified")).toBe(true);
  expect(wrapper.find(DeploymentFormBody).prop("valuesModified")).toBe(true);
});

it("triggers an upgrade when submitting the form", done => {
  const releaseName = "my-release";
  const namespace = "default";
  const appValues = "foo: bar";
  const schema = { properties: { foo: { type: "string" } } };
  const upgradeApp = jest.fn(() => true);
  const push = jest.fn();
  const wrapper = mount(
    <UpgradeForm
      {...defaultProps}
      selected={{ versions, version: versions[0], schema }}
      upgradeApp={upgradeApp}
      push={push}
      namespace={namespace}
    />,
  );
  wrapper.setState({ releaseName, appValues });
  wrapper.find("form").simulate("submit");
  expect(upgradeApp).toHaveBeenCalledWith(versions[0], releaseName, namespace, appValues, schema);
  setTimeout(() => {
    expect(push).toHaveBeenCalledWith("/apps/ns/default/my-release");
    done();
  }, 1);
});
