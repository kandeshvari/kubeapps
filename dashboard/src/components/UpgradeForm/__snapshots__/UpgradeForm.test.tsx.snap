// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the full UpgradeForm 1`] = `
<form
  className="container padding-b-bigger"
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <h2>
        my-release (my-repo/my-chart)
      </h2>
    </div>
    <div
      className="col-8"
    >
      <DeploymentFormBody
        appValues="foo: bar"
        chartID="my-repo/my-chart"
        chartVersion="1.0.0"
        fetchChartVersions={[Function]}
        getChartVersion={[Function]}
        goBack={[Function]}
        namespace="default"
        originalValues="foo: bar"
        push={[Function]}
        releaseName="my-release"
        selected={
          Object {
            "version": Object {
              "attributes": Object {
                "version": "1.2.3",
              },
              "id": "foo",
            },
            "versions": Array [
              Object {
                "attributes": Object {
                  "version": "1.2.3",
                },
                "id": "foo",
              },
            ],
          }
        }
        setValues={[Function]}
        setValuesModified={[Function]}
        valuesModified={false}
      />
    </div>
  </div>
</form>
`;
